<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2025-01-13">

<title>QC &amp; Trimming ‚Äì MCIC Bioinformatics Tutorials</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-57476271c5f0467c372e20ff5d25630e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-93a206ead970ce8646eacc93a2842d74.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/quarto-contrib/fontawesome6-1.2.0/all.min.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-1.2.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">MCIC Bioinformatics Tutorials</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-shell" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Shell</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-shell">    
        <li>
    <a class="dropdown-item" href="../shell/01_shell1.html">
 <span class="dropdown-text">Intro the Unix Shell</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell/02_shellfiles.html">
 <span class="dropdown-text">Managing files in the shell</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell/03_scripts.html">
 <span class="dropdown-text">Writing shell scripts</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell/04_cli-tools.html">
 <span class="dropdown-text">Running CLI tools</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../shell/05_bonus.html">
 <span class="dropdown-text">Advanced shell</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-osc" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">OSC</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-osc">    
        <li>
    <a class="dropdown-item" href="../osc/01_osc1.html">
 <span class="dropdown-text">Intro to OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../osc/02_vscode.html">
 <span class="dropdown-text">VS Code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../osc/03_osc2.html">
 <span class="dropdown-text">More about OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../osc/05_slurm.html">
 <span class="dropdown-text">Slurm</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../osc/06_ssh.html">
 <span class="dropdown-text">SSH</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-git" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Git</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-git">    
        <li>
    <a class="dropdown-item" href="../git/02_git1.html">
 <span class="dropdown-text">Git 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../git/03_git2.html">
 <span class="dropdown-text">Git 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../git/04_git3.html">
 <span class="dropdown-text">Git 3</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-reproducibility" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Reproducibility</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-reproducibility">    
        <li>
    <a class="dropdown-item" href="../reprod/project-org.html">
 <span class="dropdown-text">Project file organization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../reprod/markdown.html">
 <span class="dropdown-text">Markdown</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-nextflow" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Nextflow</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-nextflow">    
        <li>
    <a class="dropdown-item" href="../nf-run/01_pipelines.html">
 <span class="dropdown-text">Intro to pipelines</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../nf-run/02_nfcore.html">
 <span class="dropdown-text">Running an nf-core pipeline</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-metabar" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Metabar</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-metabar">    
        <li>
    <a class="dropdown-item" href="../metabar/01_qc-trim.html">
 <span class="dropdown-text">Read QC and trimming</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-rna-seq" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">RNA-Seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-rna-seq">    
        <li>
    <a class="dropdown-item" href="../rnaseq/01_data.html">
 <span class="dropdown-text">Data files</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../rnaseq/02_nfc-rnaseq.html">
 <span class="dropdown-text">The nf-core rnaseq pipeline</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../rnaseq/03_DE.html">
 <span class="dropdown-text">Differential expression analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../rnaseq/04_fastqc.html">
 <span class="dropdown-text">Read QC with FastQC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../rnaseq/05_trimgalore.html">
 <span class="dropdown-text">Trimming reads with TrimGalore</span></a>
  </li>  
        <li class="dropdown-header">RNA-Seq</li>
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-other" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Other</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-other">    
        <li class="dropdown-header">BACTERIAL WGS</li>
        <li>
    <a class="dropdown-item" href="../bac/01_preprocess.html">
 <span class="dropdown-text">Data preprocessing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../bac/02_assembly.html">
 <span class="dropdown-text">Genome assembly</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../bac/03_annotation.html">
 <span class="dropdown-text">Genome annotation</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" role="link" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/mcic-osu/tutorials">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://mcic-osu.github.io/tutorials/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#running-fastqc-for-1-sample" id="toc-running-fastqc-for-1-sample" class="nav-link" data-scroll-target="#running-fastqc-for-1-sample"><span class="header-section-number">1</span> Running FastQC for 1 sample</a>
  <ul class="collapse">
  <li><a href="#intro-to-fastqc" id="toc-intro-to-fastqc" class="nav-link" data-scroll-target="#intro-to-fastqc"><span class="header-section-number">1.1</span> Intro to FastQC</a></li>
  <li><a href="#our-fastq-files" id="toc-our-fastq-files" class="nav-link" data-scroll-target="#our-fastq-files"><span class="header-section-number">1.2</span> Our FASTQ files</a></li>
  <li><a href="#building-our-fastqc-command" id="toc-building-our-fastqc-command" class="nav-link" data-scroll-target="#building-our-fastqc-command"><span class="header-section-number">1.3</span> Building our FastQC command</a></li>
  <li><a href="#fastqc-output-files" id="toc-fastqc-output-files" class="nav-link" data-scroll-target="#fastqc-output-files"><span class="header-section-number">1.4</span> FastQC output files</a></li>
  </ul></li>
  <li><a href="#interpreting-fastqc-output" id="toc-interpreting-fastqc-output" class="nav-link" data-scroll-target="#interpreting-fastqc-output"><span class="header-section-number">2</span> Interpreting FastQC output</a>
  <ul class="collapse">
  <li><a href="#fastqc-html-modules" id="toc-fastqc-html-modules" class="nav-link" data-scroll-target="#fastqc-html-modules"><span class="header-section-number">2.1</span> FastQC HTML modules</a></li>
  <li><a href="#checking-your-fastqc-results" id="toc-checking-your-fastqc-results" class="nav-link" data-scroll-target="#checking-your-fastqc-results"><span class="header-section-number">2.2</span> Checking your FastQC results</a></li>
  </ul></li>
  <li><a href="#running-fastqc-for-all-samples" id="toc-running-fastqc-for-all-samples" class="nav-link" data-scroll-target="#running-fastqc-for-all-samples"><span class="header-section-number">3</span> Running FastQC for all samples</a>
  <ul class="collapse">
  <li><a href="#the-components-of-our-script" id="toc-the-components-of-our-script" class="nav-link" data-scroll-target="#the-components-of-our-script"><span class="header-section-number">3.1</span> The components of our script</a></li>
  <li><a href="#our-final-script" id="toc-our-final-script" class="nav-link" data-scroll-target="#our-final-script"><span class="header-section-number">3.2</span> Our final script</a></li>
  <li><a href="#submitting-the-script" id="toc-submitting-the-script" class="nav-link" data-scroll-target="#submitting-the-script"><span class="header-section-number">3.3</span> Submitting the script</a></li>
  <li><a href="#checking-the-output" id="toc-checking-the-output" class="nav-link" data-scroll-target="#checking-the-output"><span class="header-section-number">3.4</span> Checking the output</a></li>
  </ul></li>
  <li><a href="#summarizing-qc-results-with-multiqc" id="toc-summarizing-qc-results-with-multiqc" class="nav-link" data-scroll-target="#summarizing-qc-results-with-multiqc"><span class="header-section-number">4</span> Summarizing QC results with MultiQC</a>
  <ul class="collapse">
  <li><a href="#running-multiqc" id="toc-running-multiqc" class="nav-link" data-scroll-target="#running-multiqc"><span class="header-section-number">4.1</span> Running MultiQC</a></li>
  <li><a href="#multiqc-output" id="toc-multiqc-output" class="nav-link" data-scroll-target="#multiqc-output"><span class="header-section-number">4.2</span> MultiQC output</a></li>
  </ul></li>
  <li><a href="#cutadapt" id="toc-cutadapt" class="nav-link" data-scroll-target="#cutadapt"><span class="header-section-number">5</span> Cutadapt</a>
  <ul class="collapse">
  <li><a href="#primer-sequences" id="toc-primer-sequences" class="nav-link" data-scroll-target="#primer-sequences"><span class="header-section-number">5.1</span> Primer sequences</a></li>
  <li><a href="#building-the-cutadapt-command" id="toc-building-the-cutadapt-command" class="nav-link" data-scroll-target="#building-the-cutadapt-command"><span class="header-section-number">5.2</span> Building the Cutadapt command</a></li>
  <li><a href="#our-cutadapt-loop" id="toc-our-cutadapt-loop" class="nav-link" data-scroll-target="#our-cutadapt-loop"><span class="header-section-number">5.3</span> Our Cutadapt loop</a></li>
  <li><a href="#the-final-script" id="toc-the-final-script" class="nav-link" data-scroll-target="#the-final-script"><span class="header-section-number">5.4</span> The final script</a></li>
  <li><a href="#check-the-output" id="toc-check-the-output" class="nav-link" data-scroll-target="#check-the-output"><span class="header-section-number">5.5</span> Check the output</a></li>
  </ul></li>
  <li><a href="#bonus-content" id="toc-bonus-content" class="nav-link" data-scroll-target="#bonus-content"><span class="header-section-number">6</span> Bonus content</a>
  <ul class="collapse">
  <li><a href="#the-fastq-format" id="toc-the-fastq-format" class="nav-link" data-scroll-target="#the-fastq-format"><span class="header-section-number">6.1</span> The FASTQ format</a></li>
  <li><a href="#viewing-fastq-files" id="toc-viewing-fastq-files" class="nav-link" data-scroll-target="#viewing-fastq-files"><span class="header-section-number">6.2</span> Viewing FASTQ files</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Read QC and primer trimming</h1>
<p class="subtitle lead">With FastQC, MultiQC, and Cutadapt</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jelmer Poelstra </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 13, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<hr>
<p><br></p>
<section id="introduction" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="introduction">Introduction</h2>
<p>The first series of steps of a metabarcoding analysis workflow concerns the quality control (QC) &amp; ‚Äúpre-processing‚Äù of the raw sequence reads, which are stored in FASTQ files.</p>
<p>The QC part will leave the data untouched, whereas the pre-processing involves the removal of unwanted bits of sequence: in our case, amplicon primers. After the pre-processing step, we will still have FASTQ files, just with somewhat less content.</p>
<p>Specifically, we will go through the following steps:</p>
<ol type="1">
<li>QC with <strong>FastQC</strong></li>
<li>Summarizing FastQC results with <strong>MultiQC</strong></li>
<li>Removing primers with <strong>Cutadapt</strong></li>
</ol>
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
You should have an active VS Code session. If not, follow these steps <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Start a new VS Code session with an open terminal:</strong></p>
<ol type="1">
<li>Log in to OSC‚Äôs OnDemand portal at <a href="https://ondemand.osc.edu" class="uri">https://ondemand.osc.edu</a>.</li>
<li>In the blue top bar, select <code>Interactive Apps</code> and then near the bottom of the dropdown menu, click <code>Code Server</code>.</li>
<li>In the form that appears on a new page:
<ul>
<li>Select OSC project <code>PAS2714</code></li>
<li>The starting directory: <code>/fs/ess/PAS2714/&lt;user&gt;</code> (replace <code>&lt;user&gt;</code> with your username)</li>
<li><code>Number of hours</code>: <code>3</code></li>
<li>Click <code>Launch</code>.</li>
</ul></li>
<li>On the next page, once the top bar of the box has turned green and says <code>Runnning</code>, click <code>Connect to VS Code</code>.</li>
<li>Open a Terminal by clicking &nbsp; <i class="fa-solid fa-bars" aria-label="bars"></i> &nbsp; =&gt; <code>Terminal</code> =&gt; <code>New Terminal</code>.</li>
</ol>
</div>
</div>
</div>
<p><br></p>
</section>
<section id="running-fastqc-for-1-sample" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="running-fastqc-for-1-sample"><span class="header-section-number">1</span> Running FastQC for 1 sample</h2>
<section id="intro-to-fastqc" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="intro-to-fastqc"><span class="header-section-number">1.1</span> Intro to FastQC</h3>
<p>FastQC is a ubiquitous tools for <strong>quality control of FASTQ files</strong>. Running FastQC or a similar program is the first step in nearly any high-throughput sequencing project. FastQC is also a good introductory example of a tool with a command-line interface.</p>
<p>For each FASTQ file, FastQC outputs an <strong>HTML file</strong> that you can open in your browser with about a dozen graphs showing different QC metrics. The most important one is the <strong>per-base quality score graph</strong> shown below.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_good.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>A FastQC per-base quality score graph for files with reasonably good quality reads. The y-axis shows Phred quality scores (higher is better, see also the color-coding) and the x-axis shows the position along the read.</figcaption>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="our-fastq-files" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="our-fastq-files"><span class="header-section-number">1.2</span> Our FASTQ files</h3>
<p>Our FASTQ files contain reads from 2x300 bp (i.e.&nbsp;paired-end with 300 bp forward and 300 bp reverse reads) sequencing on an Illumina MiSeq machine.</p>
<p>Let‚Äôs take a look at our list of FASTQ files:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> data/fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 150M
-rw-r-----+ 1 jelmer PAS0471 2.0M Mar  1 17:09 NW102AB_R1.fastq.gz
-rw-r-----+ 1 jelmer PAS0471 2.6M Mar  1 17:09 NW102AB_R2.fastq.gz
-rw-r-----+ 1 jelmer PAS0471 2.3M Mar  1 17:09 NW102C_R1.fastq.gz
-rw-r-----+ 1 jelmer PAS0471 3.0M Mar  1 17:09 NW102C_R2.fastq.gz
-rw-r-----+ 1 jelmer PAS0471 1.9M Mar  1 17:09 NW103AB_R1.fastq.gz
-rw-r-----+ 1 jelmer PAS0471 2.6M Mar  1 17:09 NW103AB_R2.fastq.gz
-rw-r-----+ 1 jelmer PAS0471 2.3M Mar  1 17:09 NW103C_R1.fastq.gz
-rw-r-----+ 1 jelmer PAS0471 3.1M Mar  1 17:09 NW103C_R2.fastq.gz
# [...output truncated...]</code></pre>
<p>Note in the file listing above that:</p>
<ul>
<li>There are two files per sample: <code>_R1</code> (forward reads) and <code>_R2</code> (reverse reads). This indicates that we have data from <strong>paired-end reads</strong>, as is customary when doing amplicon metabarcoding.</li>
<li>The files all have a <code>.gz</code> extension, indicating they have been <em>compressed</em> with the gzip utility.</li>
</ul>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Don‚Äôt have the FASTQ files? Click here for instructions to copy them.
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Run the following two <code>cp</code> commands to copy the necessary data:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-rv</span> /fs/ess/PAS2714/share/data /fs/ess/PAS2714/users/<span class="va">$USER</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-rv</span> /fs/ess/PAS2714/share/results /fs/ess/PAS2714/users/<span class="va">$USER</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="building-our-fastqc-command" class="level3" data-number="1.3">
<h3 data-number="1.3" class="anchored" data-anchor-id="building-our-fastqc-command"><span class="header-section-number">1.3</span> Building our FastQC command</h3>
<p>To run FastQC, we can use the command <code>fastqc</code>.</p>
<p>If you want to analyze one of your FASTQ files with default FastQC settings, a complete FastQC command to do so would simply be <code>fastqc</code> followed by the name of the file:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this)</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> data/fastq/NW102AB_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>However, an annoying FastQC <strong>default behavior</strong> is that it writes its output files in the dir where the input files are ‚Äî in general, it‚Äôs not great practice to directly mix your primary data and results like that!</p>
<p>To figure out how we can change that behavior, first consider that many commands and bioinformatics tools alike have an <strong>option <code>-h</code> and/or <code>--help</code></strong> to print usage information to the screen. Let‚Äôs try that:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>bash: fastqc: command not found...</code></pre>
<p>However, there is a wrinkle: while FastQC is installed at OSC<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>, we have to first <strong>‚Äúload it‚Äù</strong>. The way we will do this here is with a a so-called ‚ÄúConda environment‚Äù that has FastQC installed along with the other programs we will need today.</p>
<p><strong>Here‚Äôs how we can load that Conda software environment</strong> ‚Äî we first load OSC‚Äôs (mini)conda installation, and then we can load (‚Äúactivate‚Äù) the Conda environment that I created for you:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate /fs/ess/PAS0471/jelmer/conda/mbar24</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
<section id="exercise-fastqc-help-and-output-dir" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-fastqc-help-and-output-dir"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: FastQC help and output dir</h4>
<p>Print FastQC‚Äôs help info, and figure out which option you can use to specify a custom output directory.</p>
<details>
<summary>
<em>Click for the solution</em>
</summary>
<p><code>fastqc -h</code> and <code>fastqc --help</code> will both work to show the help info.</p>
<p>You‚Äôll get quite a bit of output printed to screen, including the snippet about output directories that is reproduced below:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>  -o --outdir     Create all output files in the specified output directory.
                    Please note that this directory must exist as the program
                    will not create it.  If this option is not set then the 
                    output file for each sequence file is created in the same
                    directory as the sequence file which was processed.</code></pre>
<p>So, you can use <code>-o</code> or equivalently, <code>--outdir</code> to specify an output dir.</p>
</details>
</section>
<hr style="height:1pt; visibility:hidden;">
<p>With the added <code>--outdir</code> (or <code>-o</code>) option, let‚Äôs try to run the following FastQC command:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We'll have to first create the outdir ourselves, in this case</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/fastqc</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we run FastQC</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span> results/fastqc data/fastq/NW102AB_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>application/gzip
Started analysis of NW102AB_R1.fastq.gz
Approx 5% complete for NW102AB_R1.fastq.gz
Approx 10% complete for NW102AB_R1.fastq.gz
Approx 15% complete for NW102AB_R1.fastq.gz
[...truncated...]
Analysis complete for NW102AB_R1.fastq.gz</code></pre>
<p>Success!! üéâ</p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="fastqc-output-files" class="level3" data-number="1.4">
<h3 data-number="1.4" class="anchored" data-anchor-id="fastqc-output-files"><span class="header-section-number">1.4</span> FastQC output files</h3>
<p>Let‚Äôs take a look at the files in the output dir we specified:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 1.2M
-rw-r--r-- 1 jelmer PAS0471 241K Mar 13 14:50 NW102AB_R1_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 256K Mar 13 14:50 NW102AB_R1_fastqc.zip</code></pre>
<ul>
<li>There is a <code>.zip</code> file, which contains <strong>tables</strong> with FastQC‚Äôs data summaries.</li>
<li>There is an <code>.html</code> (HTML) file, which contains <strong>plots</strong> ‚Äî this is what we‚Äôll look at next.</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<section id="exercise-another-fastqc-run" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-another-fastqc-run"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: Another FastQC run</h4>
<p>Run FastQC for the corresponding R2 FASTQ file. Would you use the same output dir?</p>
<details>
<summary>
<em>Click for the solution</em>
</summary>
<p>Yes, it makes sense to use the same output dir, since as you could see above, the output file names have the input file identifiers in them. As such, we don‚Äôt need to worry about overwriting files, and it will be more convenient to have all results in a single dir.</p>
<p>To run FastQC for the R2 (reverse-read) file:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span> results/fastqc data/fastq/NW102AB_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Started analysis of NW102AB_R2.fastq.gz
Approx 5% complete for NW102AB_R2.fastq.gz
Approx 10% complete for NW102AB_R2.fastq.gz
Approx 15% complete for NW102AB_R2.fastq.gz
[...truncated...]
Analysis complete for NW102AB_R2.fastq.gz</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>-rw-r--r-- 1 jelmer PAS0471 241K Mar 13 14:50 NW102AB_R1_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 256K Mar 13 14:50 NW102AB_R1_fastqc.zip
-rw-r--r-- 1 jelmer PAS0471 234K Mar 13 14:53 NW102AB_R2_fastqc.html
-rw-r--r-- 1 jelmer PAS0471 244K Mar 13 14:53 NW102AB_R2_fastqc.zip</code></pre>
<p>Now, we have four files: two for each of our preceding successful FastQC runs.</p>
</details>
</section>
<p><br></p>
</section>
</section>
<section id="interpreting-fastqc-output" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="interpreting-fastqc-output"><span class="header-section-number">2</span> Interpreting FastQC output</h2>
<section id="fastqc-html-modules" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="fastqc-html-modules"><span class="header-section-number">2.1</span> FastQC HTML modules</h3>
<p>We‚Äôll now go through a couple of the FastQC plots/modules with example plots<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> with good/bad results for reference.</p>
<p>FastQC has ‚Äú<strong>pass</strong>‚Äù (checkmark in green), ‚Äú<strong>warning</strong>‚Äù (exclamation mark in orange), and ‚Äú<strong>fail</strong>‚Äù (cross in red) assessments for each module. These assessments are handy, but a ‚Äúwarning‚Äù/‚Äúfail‚Äù is not necessarily the bad news it may appear to be:</p>
<ul>
<li>Some of these modules are perhaps overly strict.</li>
<li>Some warnings and fails are easily remedied or simply not a very big deal.</li>
<li>FastQC effectively assumes that your data is derived from whole-genome shotgun sequencing ‚Äî some other types of data with different properties will therefore <em>always</em> trigger a couple of warnings and fails, but these are not meaningful. This is very much the case for metabarcoding data.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_summary2.png" class="img-fluid figure-img" style="width:30.0%"></p>
<figcaption>An example module results overview from a FastQC HTML file.<br>The green checkmarks indicate Pass, the orange exclamation marks indicate Warning, and the red crosses indicate Fail.</figcaption>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
<section id="basic-statistics" class="level4">
<h4 class="anchored" data-anchor-id="basic-statistics">Basic statistics</h4>
<p>This contains, among other things, the number of sequences (reads) and the read length range:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_mod01_basic-stats.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:85.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="per-base-quality-sequence-quality" class="level4">
<h4 class="anchored" data-anchor-id="per-base-quality-sequence-quality">Per base quality sequence quality</h4>
<p>This figure visualizes the mean per-base quality score (y-axis) along the length of the reads (x-axis). Note that:</p>
<ul>
<li>A decrease in sequence quality along the reads (from left to right) is normal.</li>
<li>R2 (reverse) reads are usually of worse quality than R1 (forward) reads.</li>
</ul>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_good.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Good / acceptable</strong></figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_bad.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Bad</strong></figcaption>
</figure>
</div>
</div>
</div>
<p>To interpret the y-axis quality scores, note the color scaling in the graphs above, and see this table for details:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Phred quality score</th>
<th>Error probability</th>
<th>Rough interpretation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>10</strong></td>
<td>1 in 10</td>
<td>terrible</td>
</tr>
<tr class="even">
<td><strong>20</strong></td>
<td>1 in 100</td>
<td>bad</td>
</tr>
<tr class="odd">
<td><strong>30</strong></td>
<td>1 in 1,000</td>
<td>good</td>
</tr>
<tr class="even">
<td><strong>40</strong></td>
<td>1 in 10,000</td>
<td>excellent</td>
</tr>
</tbody>
</table>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="per-sequence-quality-scores" class="level4">
<h4 class="anchored" data-anchor-id="per-sequence-quality-scores">Per sequence quality scores</h4>
<p>This shows the same quality scores we saw above, but now simply as a density plot of per-read averages, with the quality score now along the x-axis, and the number of reads with that quality score along the y-axis:</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_mod04_per-seq-qual_good.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Good</strong></figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_mod04_per-seq-qual_bad.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Bad</strong></figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="per-tile-sequence-quality" class="level4">
<h4 class="anchored" data-anchor-id="per-tile-sequence-quality">Per tile sequence quality</h4>
<p>This graph shows whether specific tiles (areas on the flow cell, see box below) have lower quality bases in some parts of or across all of the read.</p>
<p>Poor qualities in some but not other tiles may indicate some (transient) problems with the flow cell. For example, if a tile has poor qualities for a stretch of bases, this could indicate that there was a bubble. If such tile-based problems are severe, you can contact the sequencing facility as they may be able to re-run your data at no cost.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What is a tile? <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/tiles.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="columns">
<div class="column" style="width:43%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_mod03_per-tile_good.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Good</strong></figcaption>
</figure>
</div>
</div><div class="column" style="width:57%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastq_mod04_per-tile_moderate.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Some problems</strong></figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sequence-length-distribution" class="level4">
<h4 class="anchored" data-anchor-id="sequence-length-distribution">Sequence length distribution</h4>
<p>With Illumina sequencing, the vast majority of reads typically have nearly the same read length ‚Äî in our case, 300 bp. This graph can help you check if that is indeed the case. The module will throw a warning as soon as not all sequences are of the same length (like below), but having reads with a slightly shorter read length is normal and does not matter.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_mod08_seqlen_good.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Good</strong><br>Nearly all reads have the same read length. <br>FastQC still threw a warning.</figcaption>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastqc_mod08_seqlen_warning.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption><strong>Somewhat bad</strong><br>There are unusually many reads that are shorter than the aimed-for, 150 bp read length. Though the large majority of reads are still 150 bp.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="other-fastqc-modules" class="level4">
<h4 class="anchored" data-anchor-id="other-fastqc-modules">Other FastQC modules</h4>
<p>Another module that can help to check the overall quality of your reads is <strong>Per base N content</strong>: its line graph should typically not visibly rise above 0%.</p>
<p>The remaining modules are not that useful for metabarcoding data:</p>
<ul>
<li>Adapter content<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></li>
<li>Sequence duplication levels<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a></li>
<li>Overrepresented sequences<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a></li>
<li>Per sequence GC content<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
</section>
</section>
<section id="checking-your-fastqc-results" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="checking-your-fastqc-results"><span class="header-section-number">2.2</span> Checking your FastQC results</h3>
<p>First, you‚Äôll unfortunately have to download FastQC‚Äôs output HTML files to your computer to be able to view them:</p>
<ul>
<li>Find the FastQC HTML files in the file explorer in the VS Code side bar.</li>
<li>Right-click on one of them, click <code>Download...</code> and follow the prompt to download the file somewhere to your computer (doesn‚Äôt matter where).</li>
<li>Repeat this for the second file.</li>
<li>Then, open your computer‚Äôs file browser, find the downloaded files, and double-click on one. It should be <strong>opened in your default web browser</strong>.</li>
</ul>
<p>Alternatively, you can take a look at the FastQC output files on this website: <a href="results/fastqc/NW102AB_R1_fastqc.html">R1</a> and <a href="results/fastqc/NW102AB_R2_fastqc.html">R2</a>.</p>
<section id="exercise-interpreting-your-fastqc-results" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-interpreting-your-fastqc-results"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: Interpreting your FastQC results</h4>
<ul>
<li><p>Open the HTML file for the R1 FASTQ file and go through the modules we discussed above. Can you make sense of it? Does the data look good to you, overall?</p></li>
<li><p>Now open the HTML file for the <strong>R2</strong> FASTQ file and take a look just at the quality scores. Does it look any worse than the R1?</p></li>
</ul>
</section>
<p><br></p>
</section>
</section>
<section id="running-fastqc-for-all-samples" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="running-fastqc-for-all-samples"><span class="header-section-number">3</span> Running FastQC for all samples</h2>
<p>If we want to run FastQC for all samples, it will be much better to write a shell script and submit that as a so-called Slurm batch job, rather than running FastQC ‚Äúinteractively‚Äù like we did for the first sample.</p>
<p>This is especially true for a complete data set, which would have much larger FASTQ files and possibly more samples.</p>
<section id="the-components-of-our-script" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="the-components-of-our-script"><span class="header-section-number">3.1</span> The components of our script</h3>
<p>Here is all the code we needed for out first FastQC run:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the software</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate /fs/ess/PAS0471/jelmer/conda/mbar24</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the output dir</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/fastqc</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run FastQC</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span> results/fastqc data/fastq/NW102AB_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
<p>We will need that same code in the script, except that we will need to modify our call to <code>fastqc</code> ‚Äî we will <strong>loop over all FASTQ files</strong> as follows:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run FastQC (replacement for fastqc line above)</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> fastq_file <span class="kw">in</span> data/fastq/<span class="pp">*</span>fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">fastqc</span> <span class="at">--outdir</span> results/fastqc <span class="st">"</span><span class="va">$fastq_file</span><span class="st">"</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>We are looping over all FASTQ files with the globbing pattern <code>data/fastq/*fastq.gz</code>. The loop will run as many times as we have FASTQ files.</li>
<li>In every iteration of the loop, the <code>"$fastq_file"</code> variable will contain 1 FASTQ file name, and we will run <code>fastqc</code> for that file<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>.</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>We will be <strong>submitting this script as a batch job</strong> to the Slurm compute job scheduler. To do so, we should also add some lines at the top of the script:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --account=PAS2714</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --output=slurm-fastqc.out</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>The first line <code>#!/bin/bash</code> merely indicates that this is a shell script<a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a> rather than, say, an R or Python script.</li>
<li>The lines starting with <code>#SBATCH</code> tell Slurm some details about our compute job request (much like we did when we filled out the form to start a VS Code session):
<ul>
<li>We always need to specify an ‚Äú<code>account</code>‚Äù, i.e.&nbsp;OSC project, that should be billed.</li>
<li>The only other option (of many possible!) we will use here is to specify the <code>output</code> file: this is where any output will go that would otherwise be printed to screen, such as the FastQC progress output we saw above.</li>
</ul></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>We will also add the following line to change some shell script settings, which will cause the script to stop running if any <strong>errors</strong> occur:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Strict bash settings</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="our-final-script" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="our-final-script"><span class="header-section-number">3.2</span> Our final script</h3>
<ol type="1">
<li>Open a new file in VS Code: click <i class="fa fa-bars"></i>, then <code>File</code>, then <code>New File</code>.</li>
<li>Save the file (e.g.&nbsp;press <kbd>Ctrl/‚åò</kbd>+<kbd>S</kbd>) in your <code>scripts</code> directory as <code>fastqc.sh</code>.</li>
<li>Paste the following code in the script:</li>
</ol>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --account=PAS2714</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --output=slurm-fastqc.out</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Strict bash settings</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the software</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate /fs/ess/PAS0471/jelmer/conda/mbar24</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the output dir</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/fastqc</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Run FastQC for all FASTQ files</span></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> fastq_file <span class="kw">in</span> data/fastq/<span class="pp">*</span>fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>    <span class="ex">fastqc</span> <span class="at">--outdir</span> results/fastqc <span class="st">"</span><span class="va">$fastq_file</span><span class="st">"</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Done with script fastqc.sh"</span></span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="submitting-the-script" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="submitting-the-script"><span class="header-section-number">3.3</span> Submitting the script</h3>
<p>Submit the script to Slurm (‚Äúsubmit it to the queue‚Äù) with the <code>sbatch</code> command:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> scripts/fastqc.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Submitted batch job 27047185</code></pre>
<p>After some seconds (sometimes up to a minute or so<a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a>), the Slurm job should start and create the output file that we specified at the top of the script: <code>slurm-fastqc.out</code>.</p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="checking-the-output" class="level3" data-number="3.4">
<h3 data-number="3.4" class="anchored" data-anchor-id="checking-the-output"><span class="header-section-number">3.4</span> Checking the output</h3>
<p>In our earlier two FastQC runs, FastQC logging output (‚Äú10% complete‚Äù, etc.) was printed to screen. But because this job now runs remotely on another compute node, such output will end up in the ‚ÄúSlurm log file‚Äù whose name we specified in the script. Let‚Äôs take a look:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> slurm-fastqc.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>application/gzip
Started analysis of NW102AB_R1.fastq.gz
Approx 5% complete for NW102AB_R1.fastq.gz
Approx 15% complete for NW102AB_R1.fastq.gz
Approx 20% complete for NW102AB_R1.fastq.gz
Approx 30% complete for NW102AB_R1.fastq.gz
Approx 35% complete for NW102AB_R1.fastq.gz
Approx 45% complete for NW102AB_R1.fastq.gz
Approx 50% complete for NW102AB_R1.fastq.gz
Approx 60% complete for NW102AB_R1.fastq.gz
Approx 70% complete for NW102AB_R1.fastq.gz
Approx 75% complete for NW102AB_R1.fastq.gz
Approx 85% complete for NW102AB_R1.fastq.gz
Approx 90% complete for NW102AB_R1.fastq.gz
Analysis complete for NW102AB_R1.fastq.gz
application/gzip
Started analysis of NW102AB_R2.fastq.gz
Approx 5% complete for NW102AB_R2.fastq.gz
Approx 15% complete for NW102AB_R2.fastq.gz
Approx 20% complete for NW102AB_R2.fastq.gz
#[...output truncated...]</code></pre>
<p>That looks good, in the output I printed above we can see that FastQC ran to completion for one FASTQ file and then started a second ‚Äî and this will go on and on, for all of our 64 FASTQ files.</p>
<p>You will know that the job has successfully finished when the last few lines of the Slurm log file read ‚Äú<em>Done with script fastqc.sh</em>‚Äù and print the date and time (as per the last lines of our script!):</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span> slurm-fastqc.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Approx 75% complete for W404BC_R2.fastq.gz
Approx 85% complete for W404BC_R2.fastq.gz
Approx 90% complete for W404BC_R2.fastq.gz
Approx 95% complete for W404BC_R2.fastq.gz
Analysis complete for W404BC_R2.fastq.gz
Done with script fastqc.sh
Wed Mar  6 13:34:10 EST 2024</code></pre>
<hr style="height:1pt; visibility:hidden;">
<p>Of course, we should also check the <strong>main output files</strong> ‚Äî the HTMLs and zip files:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> results/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>NW102AB_R1_fastqc.html  NW103C_R1_fastqc.zip    NW203A_R2_fastqc.html   NW304BC_R2_fastqc.zip   NW403BC_R1_fastqc.html  W101AB_R1_fastqc.zip   W103C_R2_fastqc.html   W205A_R2_fastqc.zip    W304AB_R1_fastqc.html  W403C_R1_fastqc.zip
NW102AB_R1_fastqc.zip   NW103C_R2_fastqc.html   NW203A_R2_fastqc.zip    NW305AB_R1_fastqc.html  NW403BC_R1_fastqc.zip   W101AB_R2_fastqc.html  W103C_R2_fastqc.zip    W205BC_R1_fastqc.html  W304AB_R1_fastqc.zip   W403C_R2_fastqc.html
NW102AB_R2_fastqc.html  NW103C_R2_fastqc.zip    NW203BC_R1_fastqc.html  NW305AB_R1_fastqc.zip   NW403BC_R2_fastqc.html  W101AB_R2_fastqc.zip   W204A_R1_fastqc.html   W205BC_R1_fastqc.zip   W304AB_R2_fastqc.html  W403C_R2_fastqc.zip
NW102AB_R2_fastqc.zip   NW201AB_R1_fastqc.html  NW203BC_R1_fastqc.zip   NW305AB_R2_fastqc.html  NW403BC_R2_fastqc.zip   W101C_R1_fastqc.html   W204A_R1_fastqc.zip    W205BC_R2_fastqc.html  W304AB_R2_fastqc.zip   W404A_R1_fastqc.html
NW102C_R1_fastqc.html   NW201AB_R1_fastqc.zip   NW203BC_R2_fastqc.html  NW305AB_R2_fastqc.zip   NW404A_R1_fastqc.html   W101C_R1_fastqc.zip    W204A_R2_fastqc.html   W205BC_R2_fastqc.zip   W304C_R1_fastqc.html   W404A_R1_fastqc.zip
NW102C_R1_fastqc.zip    NW201AB_R2_fastqc.html  NW203BC_R2_fastqc.zip   NW305C_R1_fastqc.html   NW404A_R1_fastqc.zip    W101C_R2_fastqc.html   W204A_R2_fastqc.zip    W303AB_R1_fastqc.html  W304C_R1_fastqc.zip    W404A_R2_fastqc.html
NW102C_R2_fastqc.html   NW201AB_R2_fastqc.zip   NW304A_R1_fastqc.html   NW305C_R1_fastqc.zip    NW404A_R2_fastqc.html   W101C_R2_fastqc.zip    W204BC_R1_fastqc.html  W303AB_R1_fastqc.zip   W304C_R2_fastqc.html   W404A_R2_fastqc.zip
NW102C_R2_fastqc.zip    NW201C_R1_fastqc.html   NW304A_R1_fastqc.zip    NW305C_R2_fastqc.html   NW404A_R2_fastqc.zip    W103AB_R1_fastqc.html  W204BC_R1_fastqc.zip   W303AB_R2_fastqc.html  W304C_R2_fastqc.zip    W404BC_R1_fastqc.html
NW103AB_R1_fastqc.html  NW201C_R1_fastqc.zip    NW304A_R2_fastqc.html   NW305C_R2_fastqc.zip    NW404BC_R1_fastqc.html  W103AB_R1_fastqc.zip   W204BC_R2_fastqc.html  W303AB_R2_fastqc.zip   W403AB_R1_fastqc.html  W404BC_R1_fastqc.zip
NW103AB_R1_fastqc.zip   NW201C_R2_fastqc.html   NW304A_R2_fastqc.zip    NW403A_R1_fastqc.html   NW404BC_R1_fastqc.zip   W103AB_R2_fastqc.html  W204BC_R2_fastqc.zip   W303C_R1_fastqc.html   W403AB_R1_fastqc.zip   W404BC_R2_fastqc.html
NW103AB_R2_fastqc.html  NW201C_R2_fastqc.zip    NW304BC_R1_fastqc.html  NW403A_R1_fastqc.zip    NW404BC_R2_fastqc.html  W103AB_R2_fastqc.zip   W205A_R1_fastqc.html   W303C_R1_fastqc.zip    W403AB_R2_fastqc.html  W404BC_R2_fastqc.zip
NW103AB_R2_fastqc.zip   NW203A_R1_fastqc.html   NW304BC_R1_fastqc.zip   NW403A_R2_fastqc.html   NW404BC_R2_fastqc.zip   W103C_R1_fastqc.html   W205A_R1_fastqc.zip    W303C_R2_fastqc.html   W403AB_R2_fastqc.zip
NW103C_R1_fastqc.html   NW203A_R1_fastqc.zip    NW304BC_R2_fastqc.html  NW403A_R2_fastqc.zip    W101AB_R1_fastqc.html   W103C_R1_fastqc.zip    W205A_R2_fastqc.html   W303C_R2_fastqc.zip    W403C_R1_fastqc.html</code></pre>
<p>That‚Äôs a lot of files! Do we need to check all of them? Luckily not, thanks to MultiQC.</p>
<p><br></p>
</section>
</section>
<section id="summarizing-qc-results-with-multiqc" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="summarizing-qc-results-with-multiqc"><span class="header-section-number">4</span> Summarizing QC results with MultiQC</h2>
<p>Here are some challenges you may run into after running FastQC:</p>
<ul>
<li>When you have many FASTQ files, you‚Äôll generate a lot of FastQC HTML files to sort through (as we did above).</li>
<li>Even if you diligently go through each file, it‚Äôs not that easy to compare the results across samples in detail, since they are not drawn in the same graphs.</li>
</ul>
<p>MultiQC addresses these problems by aggregating FastQC results from many files, and summarizing them into a single HTML file with (still) one graph per FastQC module.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Not just for FastQC results! MultiQC can recognize and process the output of dozens of bioinformatics tools.
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>MultiQC‚Äôs graphs are also <em>interactive</em>, but here is a static example of a graph showing the mean base quality scores along the read for many FASTQ files:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_fastqc_per_base_sequence_quality_plot.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:95.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
<details>
<summary>
<i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Above, what could the two ‚Äúgroups of lines‚Äù, which diverge towards the right-hand side, represent?
</summary>
These are the files with forward (top lines, better quality) and reverse (bottom lines, worse quality) reads.
</details>
<hr style="height:1pt; visibility:hidden;">
<p>MultiQC will also create a graph comparing the number of reads across files, which can be quite useful:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_readcounts.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
<section id="running-multiqc" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="running-multiqc"><span class="header-section-number">4.1</span> Running MultiQC</h3>
<p>We will only need to run MultiQC once (because it will aggregate all FastQC results at once), and that will only take a few seconds ‚Äî therefore, we can run the command interactively without using a script.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Side note: Checking the MultiQC help <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We can check MultiQC‚Äôs help with the <code>--help</code> option:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">--help</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (Only the top part of the output is shown in the screenshot below)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_help.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
<p>As the first couple of help lines in the paler gray color explain, MultiQC will search the <code>[ANALYSIS DIRECTORY]</code>, a dir that we pass to it as an argument at the end of the command line.</p>
</div>
</div>
</div>
<p>If we tell MultiQC (command <code>multiqc</code>) about the <code>results/fastqc</code> directory like so, it should find and then aggregate all the FastQC results in there:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this - we'll complete the command in a second)</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> results/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Need to active the Conda environment?
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you don‚Äôt/no longer have the <code>mbar24</code> Conda environment active, (re-)activate it as follows:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate /fs/ess/PAS0471/jelmer/conda/mbar24</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The default output directory of MultiQC is the current working directory, so just like with FastQC, we do want to use the option for the output dir ‚Äî this is our final command and you can go ahead and execute it:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run MultiQC to summarize the FastQC results</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">--outdir</span> results/multiqc results/fastqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_stdout.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="multiqc-output" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="multiqc-output"><span class="header-section-number">4.2</span> MultiQC output</h3>
<p>Once its done, you should have the following files in the output dir:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/multiqc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 1.7M
drwxr-xr-x 2 jelmer PAS2250 4.0K Mar  13 14:57 multiqc_data
-rw-r--r-- 1 jelmer PAS2250 1.7M Mar  13 14:57 multiqc_report.html</code></pre>
<p>Go ahead and find the <code>multiqc_report.html</code> listed above in VS Code‚Äôs file browser, right-click on it and download it to your computer. Then, click on the file in your own computer to open it in your browser (i.e., just like we did with the FastQC output).</p>
<p>You can also find a copy of the <a href="results/multiqc/multiqc_report.html">MultiQC HTML output file here</a>.</p>
<section id="exercise-explore-the-multiqc-results" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-explore-the-multiqc-results"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> <strong>Exercise</strong>: Explore the MultiQC results</h4>
<p>Check for example whether patterns are consistent across samples, or if there are any outliers.</p>
</section>
<p><br></p>
</section>
</section>
<section id="cutadapt" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="cutadapt"><span class="header-section-number">5</span> Cutadapt</h2>
<p>When you prepare samples for amplicon metabarcoding, you amplify a specific region with primers, and these primers will be included in the sequences that you receive. Before we go any further, we need to <strong>remove these primer sequences</strong>, which we can do with the program Cutadapt.</p>
<p>We will write a script with a loop to run Cutadapt for all samples and submit it as a batch job like we did with FastQC.</p>
<section id="primer-sequences" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="primer-sequences"><span class="header-section-number">5.1</span> Primer sequences</h3>
<p>When we run Cutadapt, we need to tell it about our primer sequences as well as their reverse complements. We‚Äôll start by storing our particular primer sequences in variables:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Primer sequences</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="va">primer_f</span><span class="op">=</span>GTGTGYCAGCMGCCGCGGTAA</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="va">primer_r</span><span class="op">=</span>GGACTACNVGGGTWTCTAAT</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Ambiguity codes
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that the <code>Y</code>, <code>M</code>, <code>N</code>, <code>V</code> and <code>W</code> in the primer sequences are so-called ‚Äúambiguity codes‚Äù that represent multiple possible bases. For example, a <code>Y</code> represents a <code>C</code> or a <code>T</code>, and an <code>N</code> represents any of the 4 bases.</p>
</div>
</div>
<p>There are many ways of getting the reverse complement of a sequence, including manually building it up, but here we‚Äôll use a trick with the <code>tr</code> command to change each base into its complement, followed by the <code>rev</code> command to get the <em>reverse</em> complement ‚Äî for example, for the forward primer:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span> <span class="kw">|</span> <span class="fu">tr</span> ATCGYRKMBDHV TAGCRYMKVHDB <span class="kw">|</span> <span class="fu">rev</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>TTACCGCGGCKGCTGRCACAC</code></pre>
<p>Below, we‚Äôll get the reverse complement for both primers, and will store those in a variable as well using the construct <code>variable=$(command)</code><a href="#fn10" class="footnote-ref" id="fnref10" role="doc-noteref"><sup>10</sup></a>:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the reverse-complements of the primers</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="va">primer_f_rc</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span> <span class="kw">|</span> <span class="fu">tr</span> ATCGYRKMBDHV TAGCRYMKVHDB <span class="kw">|</span> <span class="fu">rev</span><span class="va">)</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="va">primer_r_rc</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="st">"</span><span class="va">$primer_r</span><span class="st">"</span> <span class="kw">|</span> <span class="fu">tr</span> ATCGYRKMBDHV TAGCRYMKVHDB <span class="kw">|</span> <span class="fu">rev</span><span class="va">)</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the sequences</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"</span><span class="va">$primer_f_rc</span><span class="st">"</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"</span><span class="va">$primer_r_rc</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>TTACCGCGGCKGCTGRCACAC
ATTAGAWACCCBNGTAGTCC</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
More about the above <code>tr | rev</code> command <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li><code>tr A T</code> changes every A to a T</li>
<li><code>tr ATC TAG</code> changes every A to a T (first character in each of the two sequences), every T to an a (second character in each of the two sequences), and every C to a G (third character in each of the two sequences).</li>
<li>Therefore, in the full <code>tr</code> command above, we list all possible bases and ambiguity codes, and then change them to their complement.</li>
<li>The <code>rev</code> command simply reverses a sequence, so that we end up with the <em>reverse</em> complement.</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> ACCT <span class="kw">|</span> <span class="fu">rev</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>TCCA</code></pre>
<hr style="height:1pt; visibility:hidden;">
</div>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="building-the-cutadapt-command" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="building-the-cutadapt-command"><span class="header-section-number">5.2</span> Building the Cutadapt command</h3>
<p>First, here is how we can tell Cutadapt about the primer sequences:</p>
<ul>
<li>With the <code>-a</code> and <code>-A</code> option we specify the primer sequences in the forward and reverse reads, respectively.</li>
<li>The forward reads should contain the forward primer at the beginning of the read. Because reads are sometimes longer than the amplicon length, the reverse primer may be present at the end of the read, but as its reverse complement. We specify this using <code>-a "$primer_f"..."$primer_r_rc"</code>.</li>
<li>Similarly, the reverse reads should contain the reverse primer at the beginning of the read, and may contain the reverse complement of the forward primer at the end of the read, which we specify using <code>-A "$primer_r"..."$primer_f_rc"</code>.</li>
</ul>
<p>All in all, our primer specification looks like this:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this - this will be part of our script)</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="dt">\</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_r_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-A</span> <span class="st">"</span><span class="va">$primer_r</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_f_rc</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Spreading commands across multiple lines with <code>\</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p>Above, I spread the command across multiple lines, which makes it a little easier to read. You can run the command exactly like that: the backslashes (<strong><code>\</code></strong>) at the end of all except the last line tell the shell that our command will continue on the next line.</p>
</div>
</div>
<p>We will also:</p>
<ul>
<li>Tell Cutadapt to only keep sequences that contain the primer<a href="#fn11" class="footnote-ref" id="fnref11" role="doc-noteref"><sup>11</sup></a>, with the <code>--trimmed-only</code> option.</li>
<li>Instruct Cutadapt to use 8 ‚Äúcores‚Äù with <code>--cores 8</code>, which can speed up the run by up to 8-fold. For our small FASTQ files, this isn‚Äôt really necessary, but for a larger dataset, that can save quite some time.</li>
</ul>
<div class="sourceCode" id="cb45"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this - this will be part of our script)</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="dt">\</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_r_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-A</span> <span class="st">"</span><span class="va">$primer_r</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_f_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--trimmed-only</span> <span class="dt">\</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--cores</span> 8</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, let‚Äôs also add the output files (<code>--output</code> for R1 and <code>--paired-output</code> for R2) and the input files (as positional arguments at the end of the command) for a single example sample. With that, we have a final example command of running Cutadapt for a single sample:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this - this will be part of our script)</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="dt">\</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_r_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-A</span> <span class="st">"</span><span class="va">$primer_r</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_f_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--trimmed-only</span> <span class="dt">\</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--cores</span> 8 <span class="dt">\</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--output</span> results/cutadapt/NW102AB_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--paired-output</span> results/cutadapt/NW102AB_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>    data/fastq/NW102AB_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>    data/fastq/NW102AB_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="our-cutadapt-loop" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="our-cutadapt-loop"><span class="header-section-number">5.3</span> Our Cutadapt loop</h3>
<p>In our script, we will run Cutadapt inside a loop, similar to how we ran FastQC. However, this case is a bit more complicated, because we need to run Cutadapt for one <strong>sample</strong> and therefore two FASTQ files at a time, rather than for one FASTQ file at a time.</p>
<p>We will do that by <strong>looping over the R1 (forward read) files only</strong>, and inside the loop, inferring the name of the R2 file:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (Don't run this - this will be part of our script)</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over the R1 files</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> R1_in <span class="kw">in</span> data/fastq/<span class="pp">*</span>R1.fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get the R2 file name with "parameter expansion"</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># This does a search-and-replace: replace "_R1" with "_R2"</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    <span class="va">R2_in</span><span class="op">=</span><span class="va">${R1_in</span><span class="op">/</span><span class="ss">_R1</span><span class="op">/</span>_R2<span class="va">}</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Report</span></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Input files: </span><span class="va">$R1_in</span><span class="st"> </span><span class="va">$R2_in</span><span class="st">"</span></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Define the output files</span></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>    <span class="va">R1_out</span><span class="op">=</span><span class="st">"</span><span class="va">$outdir</span><span class="st">"</span>/<span class="va">$(</span><span class="fu">basename</span> <span class="st">"</span><span class="va">$R1_in</span><span class="st">"</span><span class="va">)</span></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>    <span class="va">R2_out</span><span class="op">=</span><span class="st">"</span><span class="va">$outdir</span><span class="st">"</span>/<span class="va">$(</span><span class="fu">basename</span> <span class="st">"</span><span class="va">$R2_in</span><span class="st">"</span><span class="va">)</span></span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Run Cutadapt</span></span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>    <span class="ex">cutadapt</span> <span class="dt">\</span></span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">-a</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_r_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">-A</span> <span class="st">"</span><span class="va">$primer_r</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_f_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">--trimmed-only</span> <span class="dt">\</span></span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">--cores</span> 8 <span class="dt">\</span></span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">--output</span> <span class="st">"</span><span class="va">$R1_out</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">--paired-output</span> <span class="st">"</span><span class="va">$R2_out</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a>        <span class="st">"</span><span class="va">$R1_in</span><span class="st">"</span> <span class="st">"</span><span class="va">$R2_in</span><span class="st">"</span></span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="the-final-script" class="level3" data-number="5.4">
<h3 data-number="5.4" class="anchored" data-anchor-id="the-final-script"><span class="header-section-number">5.4</span> The final script</h3>
<p><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Open a new text file and save it as <code>scripts/cutadapt.sh</code>. Then paste the following code into the script:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --account=PAS2714</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --output=slurm-cutadapt.out</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=8</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Strict bash settings</span></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the software</span></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate /fs/ess/PAS0471/jelmer/conda/mbar24</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Primer sequences</span></span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a><span class="va">primer_f</span><span class="op">=</span>GTGTGYCAGCMGCCGCGGTAA</span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a><span class="va">primer_r</span><span class="op">=</span>GGACTACNVGGGTWTCTAAT</span>
<span id="cb48-17"><a href="#cb48-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-18"><a href="#cb48-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the reverse-complements of the primers</span></span>
<span id="cb48-19"><a href="#cb48-19" aria-hidden="true" tabindex="-1"></a><span class="va">primer_f_rc</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span> <span class="kw">|</span> <span class="fu">tr</span> ATCGYRKMBDHV TAGCRYMKVHDB <span class="kw">|</span> <span class="fu">rev</span><span class="va">)</span></span>
<span id="cb48-20"><a href="#cb48-20" aria-hidden="true" tabindex="-1"></a><span class="va">primer_r_rc</span><span class="op">=</span><span class="va">$(</span><span class="bu">echo</span> <span class="st">"</span><span class="va">$primer_r</span><span class="st">"</span> <span class="kw">|</span> <span class="fu">tr</span> ATCGYRKMBDHV TAGCRYMKVHDB <span class="kw">|</span> <span class="fu">rev</span><span class="va">)</span></span>
<span id="cb48-21"><a href="#cb48-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-22"><a href="#cb48-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the output dir</span></span>
<span id="cb48-23"><a href="#cb48-23" aria-hidden="true" tabindex="-1"></a><span class="va">outdir</span><span class="op">=</span>results/cutadapt</span>
<span id="cb48-24"><a href="#cb48-24" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span></span>
<span id="cb48-25"><a href="#cb48-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-26"><a href="#cb48-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over the R1 files</span></span>
<span id="cb48-27"><a href="#cb48-27" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> R1_in <span class="kw">in</span> data/fastq/<span class="pp">*</span>R1.fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb48-28"><a href="#cb48-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get the R2 file name</span></span>
<span id="cb48-29"><a href="#cb48-29" aria-hidden="true" tabindex="-1"></a>    <span class="va">R2_in</span><span class="op">=</span><span class="va">${R1_in</span><span class="op">/</span><span class="ss">_R1</span><span class="op">/</span>_R2<span class="va">}</span></span>
<span id="cb48-30"><a href="#cb48-30" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb48-31"><a href="#cb48-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Report</span></span>
<span id="cb48-32"><a href="#cb48-32" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Input files: </span><span class="va">$R1_in</span><span class="st"> </span><span class="va">$R2_in</span><span class="st">"</span></span>
<span id="cb48-33"><a href="#cb48-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb48-34"><a href="#cb48-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Define the output files</span></span>
<span id="cb48-35"><a href="#cb48-35" aria-hidden="true" tabindex="-1"></a>    <span class="va">R1_out</span><span class="op">=</span><span class="st">"</span><span class="va">$outdir</span><span class="st">"</span>/<span class="va">$(</span><span class="fu">basename</span> <span class="st">"</span><span class="va">$R1_in</span><span class="st">"</span><span class="va">)</span></span>
<span id="cb48-36"><a href="#cb48-36" aria-hidden="true" tabindex="-1"></a>    <span class="va">R2_out</span><span class="op">=</span><span class="st">"</span><span class="va">$outdir</span><span class="st">"</span>/<span class="va">$(</span><span class="fu">basename</span> <span class="st">"</span><span class="va">$R2_in</span><span class="st">"</span><span class="va">)</span></span>
<span id="cb48-37"><a href="#cb48-37" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb48-38"><a href="#cb48-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Run Cutadapt</span></span>
<span id="cb48-39"><a href="#cb48-39" aria-hidden="true" tabindex="-1"></a>    <span class="ex">cutadapt</span> <span class="dt">\</span></span>
<span id="cb48-40"><a href="#cb48-40" aria-hidden="true" tabindex="-1"></a>            <span class="at">-a</span> <span class="st">"</span><span class="va">$primer_f</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_r_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb48-41"><a href="#cb48-41" aria-hidden="true" tabindex="-1"></a>            <span class="at">-A</span> <span class="st">"</span><span class="va">$primer_r</span><span class="st">"</span>...<span class="st">"</span><span class="va">$primer_f_rc</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb48-42"><a href="#cb48-42" aria-hidden="true" tabindex="-1"></a>            <span class="at">--trimmed-only</span> <span class="dt">\</span></span>
<span id="cb48-43"><a href="#cb48-43" aria-hidden="true" tabindex="-1"></a>            <span class="at">--cores</span> 8 <span class="dt">\</span></span>
<span id="cb48-44"><a href="#cb48-44" aria-hidden="true" tabindex="-1"></a>            <span class="at">--output</span> <span class="st">"</span><span class="va">$R1_out</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb48-45"><a href="#cb48-45" aria-hidden="true" tabindex="-1"></a>            <span class="at">--paired-output</span> <span class="st">"</span><span class="va">$R2_out</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb48-46"><a href="#cb48-46" aria-hidden="true" tabindex="-1"></a>            <span class="st">"</span><span class="va">$R1_in</span><span class="st">"</span> <span class="st">"</span><span class="va">$R2_in</span><span class="st">"</span></span>
<span id="cb48-47"><a href="#cb48-47" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb48-48"><a href="#cb48-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-49"><a href="#cb48-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb48-50"><a href="#cb48-50" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Done with script cutadapt.sh"</span></span>
<span id="cb48-51"><a href="#cb48-51" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Now we <strong>submit the script</strong> as a batch job in the same way we did with the FastQC script:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> scripts/cutadapt.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Submitted batch job 27047247</code></pre>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="check-the-output" class="level3" data-number="5.5">
<h3 data-number="5.5" class="anchored" data-anchor-id="check-the-output"><span class="header-section-number">5.5</span> Check the output</h3>
<p>Once we see the ‚Äú<em>Done with script</em>‚Äù line when we use <code>tail</code> on the Slurm log file, we know the job has finished:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span> slurm-cutadapt.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>22      4       0.0     2       0 3 1
24      1       0.0     2       0 1
25      1       0.0     2       0 0 1
34      1       0.0     2       0 0 1
38      2       0.0     2       0 2
42      1       0.0     2       0 0 1
44      1       0.0     2       0 1
47      1       0.0     2       0 1
Done with script cutadapt.sh
Wed Mar  6 14:46:56 EST 2024</code></pre>
<p>Let‚Äôs use <code>less</code> to take a closer look at the logging output:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> slurm-cutadapt.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Input files: data/fastq/NW102AB_R1.fastq.gz data/fastq/NW102AB_R2.fastq.gz
This is cutadapt 4.6 with Python 3.10.13
Command line parameters: -a GTGTGYCAGCMGCCGCGGTAA...ATTAGAWACCCBNGTAGTCC -A GGACTACNVGGGTWTCTAAT...TTACCGCGGCKGCTGRCACAC --trimmed-only --cores 8 --output results/cutadapt/NW102AB_R1.fastq.gz --paired-output results/cutadapt/NW102AB_R2.fast
q.gz data/fastq/NW102AB_R1.fastq.gz data/fastq/NW102AB_R2.fastq.gz
Processing paired-end reads on 8 cores ...
Finished in 0.875 s (68.148 ¬µs/read; 0.88 M reads/minute).

=== Summary ===

Total read pairs processed:             12,844
  Read 1 with adapter:                  12,798 (99.6%)
  Read 2 with adapter:                  12,541 (97.6%)

== Read fate breakdown ==
Pairs discarded as untrimmed:              337 (2.6%)
Pairs written (passing filters):        12,507 (97.4%)
# [...output truncated...]</code></pre>
<p>We can see that Cutadapt reports the numbers and percentages of reads that contained what it calls the ‚Äúadapter‚Äù: in our case, that‚Äôs the primer. In the example above, and that is typical, the percentages are in the upper 90s.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
If you see much lower percentages here, then something is wrong, e.g.&nbsp;with the primer sequences you provided or the Cutadapt syntax you used.
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>We will use the <code>grep</code> command to print all lines that contain the information on numbers and percentages of reads with the primer sequences:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"with adapter:"</span> slurm-cutadapt.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>  Read 1 with adapter:                  12,798 (99.6%)
  Read 2 with adapter:                  12,541 (97.6%)
  Read 1 with adapter:                  14,499 (99.7%)
  Read 2 with adapter:                  14,211 (97.7%)
  Read 1 with adapter:                  12,174 (99.7%)
  Read 2 with adapter:                  11,835 (97.0%)
  Read 1 with adapter:                  15,054 (99.7%)
  Read 2 with adapter:                  14,737 (97.6%)
# [...output truncated...]</code></pre>
<p><strong>You should always take a careful look at this output</strong>, to check if there are no samples with much lower percentages: it looks like there are no such samples in this case, fortunately.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Want to quickly see the lowest % of reads with adapter? <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Use this command to extract and sort the percentages:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"with adapter:"</span> slurm-cutadapt.out <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span><span class="st">"("</span> <span class="at">-f2</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-n</span> <span class="kw">|</span> <span class="fu">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>97.0%)
97.3%)
97.5%)
97.5%)
97.5%)
97.5%)
97.6%)</code></pre>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Need to identify the samples with a specific percentage? <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Use this command to print the 7 lines preceding each match, so you can see the file names (samples) that correspond to each percentage of reads with primers ‚Äî in the box below, you‚Äôll have to scroll all the way to the right to see those file names.</p>
<p>(An alternative way, of course, would just be to scroll through the entire output file, but this is (even) more of a hassle, as Cutadapt has quite some output.)</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-B7</span> <span class="st">"with adapter:"</span> results/cutadapt/logs/slurm-cutadapt.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Command line parameters: -a GTGTGYCAGCMGCCGCGGTAA...ATTAGAWACCCBNGTAGTCC -A GGACTACNVGGGTWTCTAAT...TTACCGCGGCKGCTGRCACAC --trimmed-only --cores 8 --output results/cutadapt/NW102AB_R1.fastq.gz --paired-output results/cutadapt/NW102AB_R2.fastq.gz data/fastq/NW102AB_R1.fastq.gz data/fastq/NW102AB_R2.fastq.gz
Processing paired-end reads on 8 cores ...
Finished in 0.875 s (68.148 ¬µs/read; 0.88 M reads/minute).

=== Summary ===

Total read pairs processed:             12,844
  Read 1 with adapter:                  12,798 (99.6%)
  Read 2 with adapter:                  12,541 (97.6%)
--
Command line parameters: -a GTGTGYCAGCMGCCGCGGTAA...ATTAGAWACCCBNGTAGTCC -A GGACTACNVGGGTWTCTAAT...TTACCGCGGCKGCTGRCACAC --trimmed-only --cores 8 --output results/cutadapt/NW102C_R1.fastq.gz --paired-output results/cutadapt/NW102C_R2.fastq.gz data/fastq/NW102C_R1.fastq.gz data/fastq/NW102C_R2.fastq.gz
Processing paired-end reads on 8 cores ...
Finished in 0.801 s (55.030 ¬µs/read; 1.09 M reads/minute).

=== Summary ===

Total read pairs processed:             14,549
  Read 1 with adapter:                  14,499 (99.7%)
  Read 2 with adapter:                  14,211 (97.7%)
--
# [...output truncated...]</code></pre>
</div>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
<p>Finally, let‚Äôs check the output dir, which contains the trimmed FASTQ files we‚Äôll use in the next step of the workflow:</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> results/cutadapt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>NW102AB_R1.fastq.gz  NW103C_R1.fastq.gz   NW203A_R1.fastq.gz   NW304BC_R1.fastq.gz  NW403A_R1.fastq.gz   NW404BC_R1.fastq.gz  W103AB_R1.fastq.gz  W204BC_R1.fastq.gz  W303AB_R1.fastq.gz  W304C_R1.fastq.gz   W404A_R1.fastq.gz
NW102AB_R2.fastq.gz  NW103C_R2.fastq.gz   NW203A_R2.fastq.gz   NW304BC_R2.fastq.gz  NW403A_R2.fastq.gz   NW404BC_R2.fastq.gz  W103AB_R2.fastq.gz  W204BC_R2.fastq.gz  W303AB_R2.fastq.gz  W304C_R2.fastq.gz   W404A_R2.fastq.gz
NW102C_R1.fastq.gz   NW201AB_R1.fastq.gz  NW203BC_R1.fastq.gz  NW305AB_R1.fastq.gz  NW403BC_R1.fastq.gz  W101AB_R1.fastq.gz   W103C_R1.fastq.gz   W205A_R1.fastq.gz   W303C_R1.fastq.gz   W403AB_R1.fastq.gz  W404BC_R1.fastq.gz
NW102C_R2.fastq.gz   NW201AB_R2.fastq.gz  NW203BC_R2.fastq.gz  NW305AB_R2.fastq.gz  NW403BC_R2.fastq.gz  W101AB_R2.fastq.gz   W103C_R2.fastq.gz   W205A_R2.fastq.gz   W303C_R2.fastq.gz   W403AB_R2.fastq.gz  W404BC_R2.fastq.gz
NW103AB_R1.fastq.gz  NW201C_R1.fastq.gz   NW304A_R1.fastq.gz   NW305C_R1.fastq.gz   NW404A_R1.fastq.gz   W101C_R1.fastq.gz    W204A_R1.fastq.gz   W205BC_R1.fastq.gz  W304AB_R1.fastq.gz  W403C_R1.fastq.gz
NW103AB_R2.fastq.gz  NW201C_R2.fastq.gz   NW304A_R2.fastq.gz   NW305C_R2.fastq.gz   NW404A_R2.fastq.gz   W101C_R2.fastq.gz    W204A_R2.fastq.gz   W205BC_R2.fastq.gz  W304AB_R2.fastq.gz  W403C_R2.fastq.gz</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/cutadapt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>-rw-rw----+ 1 jelmer PAS0471 1.9M Mar  6 14:46 NW102AB_R1.fastq.gz
-rw-rw----+ 1 jelmer PAS0471 2.4M Mar  6 14:46 NW102AB_R2.fastq.gz
-rw-rw----+ 1 jelmer PAS0471 2.2M Mar  6 14:46 NW102C_R1.fastq.gz
-rw-rw----+ 1 jelmer PAS0471 2.8M Mar  6 14:46 NW102C_R2.fastq.gz
-rw-rw----+ 1 jelmer PAS0471 1.8M Mar  6 14:46 NW103AB_R1.fastq.gz
-rw-rw----+ 1 jelmer PAS0471 2.4M Mar  6 14:46 NW103AB_R2.fastq.gz
-rw-rw----+ 1 jelmer PAS0471 2.2M Mar  6 14:46 NW103C_R1.fastq.gz
-rw-rw----+ 1 jelmer PAS0471 2.9M Mar  6 14:46 NW103C_R2.fastq.gz
# [...output truncated...]</code></pre>
<p><br></p>
</section>
</section>
<section id="bonus-content" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="bonus-content"><span class="header-section-number">6</span> Bonus content</h2>
<section id="the-fastq-format" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="the-fastq-format"><span class="header-section-number">6.1</span> The FASTQ format</h3>
<p>FASTQ is a very common output format of high-throughput sequencing machines. Like most genomic data files, these are plain text files. Each sequence that is read by the sequencer (i.e., each ‚Äúread‚Äù) forms <strong>one FASTQ entry represented by four lines</strong>. The lines contain, respectively:</p>
<ol type="1">
<li>A <strong>header</strong> that starts with <code>@</code> and e.g.&nbsp;uniquely identifies the read</li>
<li>The <strong>sequence</strong> itself</li>
<li>A <strong><code>+</code></strong> (plus sign)</li>
<li>One-character <strong>quality scores</strong> for each base in the sequence</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/fastq_header.png" class="img-fluid figure-img" style="width:85.0%"></p>
<figcaption>One entry (read) in a FASTQ file covers 4 lines. <br>The header line is annotated, with some of the more useful components highlighted in red. <br>For viewing purposes, this read (at only 56 bp) is shorter than what is typical.</figcaption>
</figure>
</div>
<p>The ‚ÄúQ‚Äù in FASTQ stands for ‚Äú<em>quality</em>‚Äù, to contrast this format with FASTA, a more basic and generic sequence data format that does not include base quality scores. FASTQ files have the extension <code>.fastq</code> or <code>.fq</code>, but they are very commonly gzip-compressed, in which case their name ends in <code>.fastq.gz</code> or <code>.fq.gz</code>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
FASTQ quality scores <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The quality scores we saw in the read above represent an <strong>estimate of the error probability of the base call</strong>. Specifically, they correspond to a numeric ‚ÄúPhred‚Äù quality score (<code>Q</code>), which is a function of the estimated probability that a base call is erroneous (<code>P</code>):</p>
<blockquote class="blockquote">
<p><strong>Q = -10 * log10(P)</strong></p>
</blockquote>
<p>For some specific probabilities and their rough qualitative interpretations for Illumina data:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
<col style="width: 25%">
</colgroup>
<thead>
<tr class="header">
<th>Phred quality score</th>
<th>Error probability</th>
<th>Rough interpretation</th>
<th>ASCII character</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>10</strong></td>
<td>1 in 10</td>
<td>terrible</td>
<td><code>+</code></td>
</tr>
<tr class="even">
<td><strong>20</strong></td>
<td>1 in 100</td>
<td>bad</td>
<td><code>5</code></td>
</tr>
<tr class="odd">
<td><strong>30</strong></td>
<td>1 in 1,000</td>
<td>good</td>
<td><code>?</code></td>
</tr>
<tr class="even">
<td><strong>40</strong></td>
<td>1 in 10,000</td>
<td>excellent</td>
<td><code>?</code></td>
</tr>
</tbody>
</table>
<p>This numeric quality score is represented in FASTQ files <em>not by the number itself</em>, but by a corresponding ‚ÄúASCII character‚Äù (last column in the table). This allows for a single-character representation of each possible score ‚Äî as a consequence, <strong>each quality score character can conveniently correspond to (&amp; line up with) a base character</strong> in the read. (For your reference, <a href="https://www.drive5.com/usearch/manual/quality_score.html">here is a complete lookup table</a> ‚Äî look at the top table, ‚ÄúBASE=33‚Äù).</p>
</div>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="viewing-fastq-files" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="viewing-fastq-files"><span class="header-section-number">6.2</span> Viewing FASTQ files</h3>
<p>Next, we‚Äôll take a peak inside one of these FASTQ files.</p>
<p>The <code>head</code> command prints the first lines of a file. Let‚Äôs use it try to and print 8 lines, which should show us two reads:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 8 data/fastq/NW102AB_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>ÔøΩ
‘Ω€íÔøΩ8ÔøΩEÔøΩÔøΩ_1fÔøΩ"ÔøΩQDÔøΩJÔøΩÔøΩDÔøΩfs{ÔøΩÔøΩÔøΩÔøΩYkÔøΩÔøΩÔøΩÔøΩdÔøΩÔøΩ*ÔøΩÔøΩ
|ÔøΩÔøΩxÔøΩÔøΩÔøΩlﬁ¥ÔøΩjÔøΩNÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩ?ÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩŸîÔøΩbUsÔøΩNgÔøΩ«¨ÔøΩÔøΩÔøΩi;_ÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩ|&lt;ÔøΩvÔøΩÔøΩÔøΩÔøΩ3ÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩ|ÔøΩÔøΩÔøΩ€ßÔøΩÔøΩ3ƒêHy∆ïÔøΩbIŒüDÔøΩ%ÔøΩÔøΩÔøΩÔøΩSr#~ÔøΩÔøΩ7ÔøΩÔøΩŒΩÔøΩÔøΩ1yÔøΩAi,4
w\]"bÔøΩ#QÔøΩÔøΩÔøΩÔøΩ8ÔøΩÔøΩ+[eÔøΩ3dÔøΩ4HÔøΩÔøΩÔøΩÃíÔøΩlÔøΩ9LVMXÔøΩÔøΩU*ÔøΩMÔøΩÔøΩÔøΩÔøΩ_?ÔøΩÔøΩÔøΩ\["ÔøΩÔøΩ7ÔøΩs\&lt;_ÔøΩÔøΩÔøΩ:ÔøΩ$ÔøΩÔøΩÔøΩNÔøΩÔøΩvÔøΩ}^ÔøΩÔøΩÔøΩÔøΩswÔøΩ|ÔøΩn;&lt;ÔøΩ&lt;ÔøΩoPÔøΩÔøΩÔøΩÔøΩ
iÔøΩÔøΩkÔøΩÔøΩqÔøΩ÷∞(GÔøΩœ´ÔøΩÔøΩLÔøΩ^ÔøΩÔøΩ=ÔøΩÔøΩ&lt;ÔøΩÔøΩÔøΩKÔøΩÔøΩjÔøΩ_/ÔøΩ[€≠VÔøΩns:ÔøΩÔøΩUÔøΩÔøΩGÔøΩzÔøΩ›éÔøΩjÔøΩÔøΩÔøΩÔøΩ&amp;ÔøΩÔøΩ~ÔøΩFÔøΩÔøΩŸ§ZNÔøΩ'ÔøΩÔøΩr2z}ÔøΩf\#ÔøΩÔøΩ:ÔøΩ9$ÔøΩÔøΩÔøΩÔøΩÔøΩHÔøΩ›ÇÔøΩ"ÔøΩ@MÔøΩÔøΩÔøΩÔøΩHÔøΩCÔøΩ
ÔøΩ0ÔøΩppÔøΩÔøΩÔøΩ1ÔøΩOÔøΩÔøΩIÔøΩHÔøΩPÎêÑÔøΩ.»¢eÔøΩÔøΩQÔøΩ&gt;ÔøΩÔøΩÔøΩ
ÔøΩ'ÔøΩ;@D8ÔøΩÔøΩÔøΩ#ÔøΩÔøΩStÔøΩ7kÔøΩgÔøΩÔøΩ|ÔøΩA‰âªÔøΩÔøΩÔøΩ_ÔøΩÔøΩÔøΩdÔøΩ_cÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩa\ÔøΩ|ÔøΩ_ÔøΩmnÔøΩ]ÔøΩ9NÔøΩÔøΩÔøΩÔøΩÔøΩÔøΩlÔøΩŸ¢ZNÔøΩcÔøΩ9uÔøΩÔøΩÔøΩÔøΩÔøΩnÔøΩÔøΩnÔøΩ`ÔøΩÔøΩ
"gÕ∫ÔøΩ
    ÔøΩÔøΩÔøΩHÔøΩ?2@ÔøΩFCÔøΩS$nÔøΩÔøΩÔøΩ‘íhÔøΩ       n‘ôjÔøΩÔøΩÊúõÔøΩÔøΩf      ÔøΩ?N@ÔøΩCzUlTÔøΩ&amp;ÔøΩhÔøΩPt!ÔøΩr|ÔøΩÔøΩ9~)ÔøΩÔøΩÔøΩeÔøΩAÔøΩ77ÔøΩh{ÔøΩÔøΩ~ÔøΩÔøΩ     ÔøΩÔøΩ
# [...output truncated...]</code></pre>
<details>
<summary>
Ouch! üò≥ What went wrong here? <em>(Click for the solution)</em>
</summary>
What happened here is that we are directly seeing the contents of the <em>compressed</em> file, which is simply not human-readable.
</details>
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
No need to decompress
</div>
</div>
<div class="callout-body-container callout-body">
<p>To get around the problem we just encountered with <code>head</code>, we might be inclined to <strong>uncompress</strong> these files, which we could do with the <strong><code>gunzip</code> command</strong>. However, uncompressed files take up several times as much disk storage space as compressed ones. Fortunately, we don‚Äôt need to decompress them:</p>
<ul>
<li>Almost any bioinformatics tool will accept compressed FASTQ files.</li>
<li>We can still view these files in compressed form, as shown below.</li>
</ul>
</div>
</div>
<p>Instead, we‚Äôll use the <code>less</code> command, which will automatically display gzip-compressed files in human-readable form:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> <span class="at">-S</span> data/fastq/NW102AB_R1.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>@M02815:77:000000000-KPK85:1:2101:3678:10660 1:N:0:CCTAAGAC+TTCTAGCT
CGAGCAATCCACTCGAGTGCCAGCAGCCGCAGTAATACGGAGGGTGCGAGCGTTGTCCGGAATCACTGGGCGTAAAGGGCGCGTAGGCGGCGCGGATAGTCGGCGGTGAAAGCCCGGAGCTCAACTCCGGGTCGGCCGTCGATACTTCCGGGCTTGAGCACTGTAGAGGCAGATGGAATTCCGGGTGTAGCGGTGGAATGCGTAGAGATCCGGAAGAACACCGGTGGCGAAGGCGGTCTGCTGGGCAGTTGCTGACGCTGATGCGCGACAGCGTGGGGAGCAAACAGGATTAGATACC
+
CCCCCGGGGGGGGGGGGGGFGGGGGGGGGG+CFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGDGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGDGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGGGGGGGGEGGGGGGGGGGGGGGGGGGGGDGGGGGGGGGGGGGGFGGFGFFFFEBFFGFFFDGFGFGBFGFGFGFFFF6?FFFGBF?FBFFF
@M02815:77:000000000-KPK85:1:2108:2535:14400 1:N:0:CCTAAGAC+TTCTAGCT
CGAGCAATCCACTCGAGTGTCAGCCGCCGCGGTAATACAGAGGTCCCGAGCGTTGTTCGGATTCATTGGGCGTAAAGGGTGCGTAGGCGGCGGGGAAAGTCTGATGTGAAATCCTGGGGCTCAACCCTGGAACTGCATTGGATACTTCCTTGCTAGAGTACTGGAGAGGAAACTGGAATTTACGGTGTAGCAGTGAAATGCGTAGAGATCGTAAGGAAGACCAGTGGCGAAGGCGAGTTTCTGGACAGTTACTGACGCTGAGGCACGAAGGCCAGGGGAGCAAACGGGATTAGATACC
+
CCCCCCGFGFGGGC-FFFGFGFFGGDFFGGGGGECGEGGAEGGGGGGGFGGDGG7CFFGGDCCFGGFCF8FGGGGGGCEGDGGGGGCGGGGGGDEGGGGBFGGDFGGGDG&lt;DFGGGGCEGGGD:FFGGGGFFGFGGFFFFGGGFGGCFGGFGGGGG9CGCGGGG7FGGC:FFGGGGGFGG&lt;?FCGGGGGGGGGGG9CG&lt;ACC?EG5CFGGGGF8CCCC:C@FGCFGGGGGC58=EEG8??77:9@:&lt;3A&gt;7AGFGGGGC?DFC?5&lt;5&gt;&gt;BGGGFGGGGG&gt;4?C42::3:DG=&gt;&lt;&lt;*)*</code></pre>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<code>less -S</code> suppresses line-wrapping: lines in the file will not be ‚Äúwrapped‚Äù across multiple lines
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">

</div>
</div>
</div>
<section id="exercise-explore-the-file-with-less" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-explore-the-file-with-less"><strong>Exercise</strong>: Explore the file with <code>less</code></h4>
<p><code>less</code> doesn‚Äôt print stuff to screen but instead opens it in a ‚Äúpager‚Äù. After running the command above, you should be viewing the file inside the <code>less</code> pager.</p>
<p>You can move around in the file in several ways: by scrolling with your mouse, with up and down arrows, or, if you have them, <kbd>PgUp</kbd> and <kbd>PgDn</kbd> keys (also, <kbd>u</kbd> will move up half a page and <kbd>d</kbd> down half a page).</p>
<p>Notice you won‚Äôt get your shell prompt back until you <strong>press</strong> <kbd>q</kbd> to quit <code>less</code>.</p>
</section>
<p><br><br></p>


</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>For a full list of installed software at OSC: <a href="https://www.osc.edu/resources/available_software/software_list" class="uri">https://www.osc.edu/resources/available_software/software_list</a><a href="#fnref1" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn2"><p> Attribution: Some of the FastQC example plots were taken from <a href="https://rtsf.natsci.msu.edu/sites/_rtsf/assets/File/FastQC_TutorialAndFAQ_080717.pdf">here</a>.<a href="#fnref2" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn3"><p>Checks for adapters at the ends of reads. Since we have to remove primers anyway, any adapters past the primers will be automatically removed.<a href="#fnref3" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn4"><p> Will throw a Fail but this is not meaningful here: metabarcoding data has many duplicate sequences by design.<a href="#fnref4" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn5"><p>As with the previous module, this will throw a Fail but this is not meaningful here: metabarcoding data has many duplicate sequences by design.<a href="#fnref5" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn6"><p>Useful in a whole-genome sequencing context or to detect contamination.<a href="#fnref6" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn7"><p>Therefore, our FastQC analysis will run sequentially (1-by-1) for each file, not in parallel.<a href="#fnref7" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn8"><p> Using the shell language Bash, specifically<a href="#fnref8" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn9"><p> And very large jobs can sometimes take hours to start, but our jobs are small so that should not happen.<a href="#fnref9" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn10"><p>This is called ‚Äúcommand substitution‚Äù.<a href="#fnref10" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn11"><p> This is not the default: Cutadapt is even more commonly used to remove adapters, and then this doesn‚Äôt apply<a href="#fnref11" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/mcic-osu\.github\.io\/tutorials");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2024-2025, Jelmer Poelstra</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/mcic-osu/tutorials">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>